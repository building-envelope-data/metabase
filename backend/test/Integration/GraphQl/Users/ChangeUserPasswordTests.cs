using System.Diagnostics.CodeAnalysis;
using System.Collections.Generic;
using System.IO;
using System.Threading.Tasks;
using FluentAssertions;
using Snapshooter.NUnit;
using NUnit.Framework;

namespace Metabase.Tests.Integration.GraphQl.Users;

[TestFixture]
public sealed class ChangeUserPasswordTests
    : UserIntegrationTests
{
    [Test]
    [SuppressMessage("Naming", "CA1707")]
    public async Task ValidData_ChangesUserPassword()
    {
        // Arrange
        const string email = "john.doe@ise.fraunhofer.de";
        const string password = "aaaAAA123$!@";
        await RegisterAndConfirmAndLoginUser(
            email: email,
            password: password
        ).ConfigureAwait(false);
        const string newPassword = "new" + password;
        // Act
        var response = await ChangeUserPassword(
            password,
            newPassword
        ).ConfigureAwait(false);
        // Assert
        Snapshot.Match(
            response,
            matchOptions => matchOptions.Assert(fieldOptions =>
                fieldOptions.Field<string>("data.changeUserPassword.user.id").Should().NotBeNullOrWhiteSpace()
            )
        );
        await LoginUser(
            email,
            newPassword
        ).ConfigureAwait(false);
    }

    [Test]
    [SuppressMessage("Naming", "CA1707")]
    public async Task NonLoggedInUser_IsAuthenticationError()
    {
        // Arrange
        const string email = "john.doe@ise.fraunhofer.de";
        const string password = "aaaAAA123$!@";
        await RegisterAndConfirmUser(
            email: email,
            password: password
        ).ConfigureAwait(false);
        const string newPassword = "new" + password;
        // Act
        var response = await UnsuccessfullyQueryGraphQlContentAsString(
            File.ReadAllText("Integration/GraphQl/Users/ChangeUserPassword.graphql"),
            variables: new Dictionary<string, object?>
            {
                ["currentPassword"] = password,
                ["newPassword"] = newPassword,
                ["newPasswordConfirmation"] = newPassword
            }
        ).ConfigureAwait(false);
        // Assert
        Snapshot.Match(response);
        await LoginUser(
            email,
            password
        ).ConfigureAwait(false);
    }

    [Test]
    [SuppressMessage("Naming", "CA1707")]
    public async Task UnconfirmedUser_IsError()
    {
        // Arrange
        const string email = "john.doe@ise.fraunhofer.de";
        const string password = "aaaAAA123$!@";
        await RegisterAndConfirmAndLoginUser(
            email: email,
            password: password
        ).ConfigureAwait(false);
        // Act
        var response = await ChangeUserPassword(
            password,
            "new" + password
        ).ConfigureAwait(false);
        // Assert
        Snapshot.Match(
            response,
            matchOptions => matchOptions
                .Assert(fieldOptions =>
                    fieldOptions.Field<string>("data.changeUserPassword.user.id").Should().NotBeNullOrWhiteSpace()
                )
        );
    }

    [Test]
    [SuppressMessage("Naming", "CA1707")]
    public async Task PasswordConfirmationMismatch_IsUserError()
    {
        // Arrange
        const string email = "john.doe@ise.fraunhofer.de";
        const string password = "aaaAAA123$!@";
        await RegisterAndConfirmAndLoginUser(
            email: email,
            password: password
        ).ConfigureAwait(false);
        // Act
        var response = await ChangeUserPassword(
            password,
            "new" + password,
            "other" + password
        ).ConfigureAwait(false);
        // Assert
        Snapshot.Match(
            response,
            matchOptions => matchOptions.Assert(fieldOptions =>
                fieldOptions.Field<string>("data.changeUserPassword.user.id").Should().NotBeNullOrWhiteSpace()
            )
        );
        await LoginUser(
            email,
            password
        ).ConfigureAwait(false);
    }

    [Test]
    [SuppressMessage("Naming", "CA1707")]
    public async Task PasswordRequiresDigit_IsUserError()
    {
        // Arrange
        const string email = "john.doe@ise.fraunhofer.de";
        const string password = "aaaAAA123$!@";
        await RegisterAndConfirmAndLoginUser(
            email: email,
            password: password
        ).ConfigureAwait(false);
        // Act
        var response = await ChangeUserPassword(
            password,
            "aabb@$CCDD"
        ).ConfigureAwait(false);
        // Assert
        Snapshot.Match(
            response,
            matchOptions => matchOptions.Assert(fieldOptions =>
                fieldOptions.Field<string>("data.changeUserPassword.user.id").Should().NotBeNullOrWhiteSpace()
            )
        );
        await LoginUser(
            email,
            password
        ).ConfigureAwait(false);
    }

    [Test]
    [SuppressMessage("Naming", "CA1707")]
    public async Task PasswordRequiresLower_IsUserError()
    {
        // Arrange
        const string email = "john.doe@ise.fraunhofer.de";
        const string password = "aaaAAA123$!@";
        await RegisterAndConfirmAndLoginUser(
            email: email,
            password: password
        ).ConfigureAwait(false);
        // Act
        var response = await ChangeUserPassword(
            password,
            "AABB@$567"
        ).ConfigureAwait(false);
        // Assert
        Snapshot.Match(
            response,
            matchOptions => matchOptions.Assert(fieldOptions =>
                fieldOptions.Field<string>("data.changeUserPassword.user.id").Should().NotBeNullOrWhiteSpace()
            )
        );
        await LoginUser(
            email,
            password
        ).ConfigureAwait(false);
    }

    [Test]
    [SuppressMessage("Naming", "CA1707")]
    public async Task PasswordRequiresNonAlphanumeric_IsUserError()
    {
        // Arrange
        const string email = "john.doe@ise.fraunhofer.de";
        const string password = "aaaAAA123$!@";
        await RegisterAndConfirmAndLoginUser(
            email: email,
            password: password
        ).ConfigureAwait(false);
        // Act
        var response = await ChangeUserPassword(
            password,
            "aaBBccDDeeFF123"
        ).ConfigureAwait(false);
        // Assert
        Snapshot.Match(
            response,
            matchOptions => matchOptions.Assert(fieldOptions =>
                fieldOptions.Field<string>("data.changeUserPassword.user.id").Should().NotBeNullOrWhiteSpace()
            )
        );
        await LoginUser(
            email,
            password
        ).ConfigureAwait(false);
    }

    [Test]
    [SuppressMessage("Naming", "CA1707")]
    public async Task PasswordRequiresUpper_IsUserError()
    {
        // Arrange
        const string email = "john.doe@ise.fraunhofer.de";
        const string password = "aaaAAA123$!@";
        await RegisterAndConfirmAndLoginUser(
            email: email,
            password: password
        ).ConfigureAwait(false);
        // Act
        var response = await ChangeUserPassword(
            password,
            "aabb@$567"
        ).ConfigureAwait(false);
        // Assert
        Snapshot.Match(
            response,
            matchOptions => matchOptions.Assert(fieldOptions =>
                fieldOptions.Field<string>("data.changeUserPassword.user.id").Should().NotBeNullOrWhiteSpace()
            )
        );
        await LoginUser(
            email,
            password
        ).ConfigureAwait(false);
    }

    [Test]
    [SuppressMessage("Naming", "CA1707")]
    public async Task PasswordTooShort_IsUserError()
    {
        // Arrange
        const string email = "john.doe@ise.fraunhofer.de";
        const string password = "aaaAAA123$!@";
        await RegisterAndConfirmAndLoginUser(
            email: email,
            password: password
        ).ConfigureAwait(false);
        // Act
        var response = await ChangeUserPassword(
            password,
            "aA@$567"
        ).ConfigureAwait(false);
        // Assert
        Snapshot.Match(
            response,
            matchOptions => matchOptions.Assert(fieldOptions =>
                fieldOptions.Field<string>("data.changeUserPassword.user.id").Should().NotBeNullOrWhiteSpace()
            )
        );
        await LoginUser(
            email,
            password
        ).ConfigureAwait(false);
    }
}