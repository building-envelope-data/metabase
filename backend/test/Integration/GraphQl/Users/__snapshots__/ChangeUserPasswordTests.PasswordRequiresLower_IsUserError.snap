{"data":{"changeUserPassword":{"user":{"id":"VXNlcgpnNzExZTcyNTA5MTQ3NDUzNTg3NGIyYzYwOTY0NDQ4MTM=","email":"john.doe@ise.fraunhofer.de"},"errors":[{"code":"PASSWORD_REQUIRES_LOWER","message":"Passwords must have at least one lowercase (\u0027a\u0027-\u0027z\u0027).","path":["input","newPassword"]}]}}}
